src/clientdlg.cpp

```diff
@@ -750,11 +750,12 @@ void CClientDlg::OnConnectDisconBut()
     // the connect/disconnect button implements a toggle functionality
```

Comment:
This should be implemented by switching the `connect` setting for the signal handler for the button, rather than having one signal handler with an if that consults the state of the client.

(Button life-cycle:
- default: onClickConnect (connects ConnectDlg onClick handler to onConnectDlgConnect and) launches ConnectDlg
- onConnectDlgConnect _disables_ the button and asks the client to connect
- onConnect switches the onClick handler to onClickDisconnect and enables the button
- onConnectFailed just enables the button (keeping the onClickConnect handler)
- onClickDisconnect _disables_ the button and asks the client to disconnect

Reply:
For JSON-RPC, the equivalent is
- (get a list of servers from somewhere)
- Send a Client JSON-RPC request to connect to a server - get an "ACK" back to say message received
- At some asynchronous time in the future, receive a Client JSON-RPC push of either a successful connect or failed connect
- (assuming connected) Send a Client JSON-RPC request to disconnect the current connection - get an "ACK" back to say message received
- At some asynchronous time in the future, receive a Client JSON-RPC push when the client decides it has disconnected

Note that in both the UI and JSON-RPC cases, the actual change in client state is asynchronous.  For disconnect, it's possible (given that the client might not actually be connected), that _nothing will happen_.

Now, for that "nothing happened" case, we've a potential problem: the UI/JSON-RPC "thought" we were connected and asked to disconnect.  Is asking the client to disconnect all that's needed or should the UI/JSON-RPC _wait_ for confirmation of the disconnect?  If it _should wait_, we then _MUST_ send that "disconnected" message back, whether or not we were connected.

Reply:
Yes, this should be changed. But in a follow up PR

Reply:
https://github.com/jamulussoftware/jamulus/issues/3401